File: clients/php/Model/Action.php
<?php

namespace Model {
    require_once 'Model/ActionType.php';
    require_once 'Stream.php';

    /**
     * Action unit is currently performing
     */
    class Action
    {
        /**
         * Tick when the action will be finished
         */
        public int $finishTick;
        /**
         * Type of the action
         */
        public int $actionType;
    
        function __construct(int $finishTick, int $actionType)
        {
            $this->finishTick = $finishTick;
            $this->actionType = $actionType;
        }
    
        /**
         * Read Action from input stream
         */
        public static function readFrom(\InputStream $stream): Action
        {
            $finishTick = $stream->readInt32();
            $actionType = \Model\ActionType::readFrom($stream);
            return new Action($finishTick, $actionType);
        }
        
        /**
         * Write Action to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32($this->finishTick);
            $stream->writeInt32($this->actionType);
        }
    }
}

File: clients/go/model/action.go
package model

import "fmt"
import "io"
import . "ai_cup_22/stream"

// Action unit is currently performing
type Action struct {
    // Tick when the action will be finished
    FinishTick int32
    // Type of the action
    ActionType ActionType
}

func NewAction(finishTick int32, actionType ActionType) Action {
    return Action {
        FinishTick: finishTick,
        ActionType: actionType,
    }
}

// Read Action from reader
func ReadAction(reader io.Reader) Action {
    var finishTick int32
    finishTick = ReadInt32(reader)
    var actionType ActionType
    actionType = ReadActionType(reader)
    return Action {
        FinishTick: finishTick,
        ActionType: actionType,
    }
}

// Write Action to writer
func (action Action) Write(writer io.Writer) {
    finishTick := action.FinishTick
    WriteInt32(writer, finishTick)
    actionType := action.ActionType
    WriteInt32(writer, int32(actionType))
}

// Get string representation of Action
func (action Action) String() string {
    stringResult := "{ "
    stringResult += "FinishTick: "
    finishTick := action.FinishTick
    stringResult += fmt.Sprint(finishTick)
    stringResult += ", "
    stringResult += "ActionType: "
    actionType := action.ActionType
    stringResult += ActionTypeToString(actionType)
    stringResult += " }"
    return stringResult
}